gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java

```diff
@@ -852,18 +857,50 @@ public void rollback(final Object callbackArg) {
     }
     // lock the TXRegionStates against GII and TXState
     lockTXRSAndTXState();
-
+    TransactionException cleanEx = null;
     final TransactionObserver observer = getObserver();
     if (observer != null) {
       observer.duringIndividualRollback(this.proxy, callbackArg);
       try {
         cleanup(false, observer);
+
       } finally {
         observer.afterIndividualRollback(this.proxy, callbackArg);
       }
+
+      try {
+        rollBackUncommittedEntries();
+      } catch (Throwable t) {
+        Error err;
+        if (t instanceof Error && SystemFailure.isJVMFailureError(
+            err = (Error)t)) {
+          SystemFailure.initiateFailure(err);
+          // If this ever returns, rethrow the error. We're poisoned
+          // now, so don't let this thread continue.
+          throw err;
+        }
+        cleanEx = processCleanupException(t, cleanEx);
+      } finally {
+
```

Comment:
If finally block is empty why do we need it? Remove it. I think the compiler will remove it. But anyways.
This code and the below can be put in a single method. Exact same code.